### ISSUE TITLE
Issue when exporting yolov8 pytorch model to tflite . 

### ISSUE THREAD

**Issue by bhava07** [ultralytics/ultralytics#9089]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey I wanted to run inference on yolov8 model . I have installed Ultralytics  .
**yolo predict model=yolov8n.pt source='https://ultralytics.com/images/zidane.jpg'**.  i was able to get proper classes and bounding boxes for the above command 
**yolo predict task=detect model=yolov8n_saved_model/yolov8n_float32.tflite imgsz=640 source= "image.jpg"**
i was able to get good results for the above command as well .
i used **yolo export model=yolov8n.pt format=tflite**  to convert the pytorch model to tflite . after conversion i wanted to run inference on the tflite model and used the below code 

import numpy as np
from PIL import Image, ImageDraw, ImageFont
import tensorflow as tf
import time
import cv2
import matplotlib.pyplot as plt
model_path = 'yolov8n_float32.tflite'
interpreter = tf.lite.Interpreter(model_path=model_path)
interpreter.allocate_tensors()

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Obtain the height and width of the corresponding image from the input tensor
image_height = input_details[0]['shape'][1] # 640
image_width = input_details[0]['shape'][2] # 640
print(input_details)
print(image_height,image_width)
# print(input_details[0]['shape'][0])
# Image Preparation
image_name = "zidane.jpg"
image = Image.open(image_name)
image_resized = image.resize((image_width, image_height))

image_np = np.array(image_resized, dtype=np.float32)

image_np = np.expand_dims(image_np, axis=0)

interpreter.set_tensor(input_details[0]['index'], image_np)

start = time.time()
interpreter.invoke()
print(f'run time：{time.time() - start:.2f}s')

output = interpreter.get_tensor(output_details[0]['index'])
output = output[0]
output = output.T

boxes_xywh = output[..., :4] #Get coordinates of bounding box, first 4 columns of output tensor
scores = np.max(output[..., 5:], axis=1) #Get score value, 5th column of output tensor
classes = np.argmax(output[..., 5:], axis=1)

print("boxes_xywh",boxes_xywh)
print("scores",scores)
print("classes",classes)
# Threshold Setting
threshold = 0.3

img = ImageDraw.Draw(image_resized)

print("boxes_xywh",boxes_xywh)
print("scores",scores)
print("classes",classes)

threshold = 0.9

draw = ImageDraw.Draw(image_resized)
#image_np_with_boxes = Image.new("RGB", (image_width, image_height))
#draw = ImageDraw.Draw(image_np_with_boxes)
filtered_indices = [i for i, score in enumerate(scores) if score >= threshold]
filtered_boxes = [boxes_xywh[i] for i in filtered_indices]
filtered_scores = [scores[i] for i in filtered_indices]
filtered_classes = [classes[i] for i in filtered_indices]

boxes_cv = []
for box in filtered_boxes:
    x_center, y_center, width, height = box
    x1 = int((x_center - width / 2) * image_width)
    y1 = int((y_center - height / 2) * image_height)
    x2 = int((x_center + width / 2) * image_width)
    y2 = int((y_center + height / 2) * image_height)
    boxes_cv.append([x1, y1, x2, y2])

# Convert scores to numpy array
scores_np = np.array(filtered_scores)

# Perform non-maximum suppression using OpenCV
indices = cv2.dnn.NMSBoxes(boxes_cv, scores_np, threshold, 0.3)

# Ensure indices is not empty before iterating
if len(indices) > 0:
    # Draw bounding boxes and text for picked indices
    for i in indices.flatten():
        box = boxes_cv[i]
        score = filtered_scores[i]
        cls = filtered_classes[i]

        x1, y1, x2, y2 = box
        draw.rectangle([x1, y1, x2, y2], outline="red", width=2)
        text = f"Class: {cls}, Score: {score:.2f}"
        draw.text((x1, y1), text, fill="red")
image_resized.save("out1.jpg")

the classes were predicted as boxes_xywh [[0.02931743 0.00532795 0.05725956 0.00414374]
 [0.03675719 0.00487283 0.07186553 0.00675204]
 [0.04787004 0.0072808  0.04770713 0.01300208]
 ...
 [0.8887477  0.8063591  0.22610593 0.3902508 ]
 [0.88642234 0.8032049  0.22692204 0.3939045 ]
 [0.88651955 0.8023272  0.23032016 0.39808714]]
scores [1.0359047e-01 9.5717758e-01 9.9228358e-01 ... 3.9056680e-04 1.6754486e-04
 1.5695423e-03]
classes [44 44 44 ... 59 59 59]

which are incorrect the expected output for class is [0 0 27]
 
i tried for both float32 , float16 and int8 and the issue persists for all the models . Is the issue with conversion of the model or the post processing . 
Please help me out Thanks 

![out1](https://github.com/ultralytics/ultralytics/assets/94779682/e816765e-e18e-4ef9-ad57-c9487508d560)
the output for tflite model is above . 

the expected output when used yolo predict by ultralytics is 
![zidane (1)](https://github.com/ultralytics/ultralytics/assets/94779682/aa7b8f60-81bb-4e3b-83db-484fefdddfaa)

thank you .

**Comment by glenn-jocher** [comment]
@bhava07 hello! Thank you for providing a detailed overview of your process and the issue you're experiencing. From your description and code, it appears that the issue might be related to how the outputs of the TFLite model are being post-processed. In TFLite, the model's output tensors might not match those of the PyTorch model directly, especially regarding the ordering and scaling of the coordinates, scores, and class predictions.

Here are a couple of suggestions:
1. **Verify output tensor structure**: Ensure the structure and order of output tensors from the TFLite model match your expectations and post-processing code. It's useful to print and inspect the `output_details` to understand the output tensor's shape and data type.
2. **Adjust post-processing**: The scales and offsets applied during post-processing might need adjustments to convert model outputs to actual bounding box coordinates and scores properly. The difference in expected vs. actual classes suggests there might be an issue with how classes are determined (e.g., incorrect index due to one-hot encoding).

For a quick test, you could directly use the highest scoring box and class to verify if at least one detection aligns with expectations:

```python
# Simplified example for testing
max_score_index = np.argmax(scores)
if scores[max_score_index] >= threshold:
    box = boxes_xywh[max_score_index]
    cls = classes[max_score_index]
    score = scores[max_score_index]
    
    # Convert and draw the box here as done in your loop
```
This simplification helps isolate whether the issue is with all detections or specific to how scores and classes are being filtered.

Remember, the way TFLite handles inputs and outputs can be quite different, especially around data types, so ensuring your input image is correctly prepared (normalized if required, correct dtype) is also crucial.

If the issue persists, it might be useful to revisit the export process and double-check if all export optimizations and quantizations are compatible with your post-processing steps. Each TFLite conversion option (`float16`, `int8`) can introduce nuances in model output that may need adjustments on the inference code.

If you continue to experience difficulties or have further questions, please feel free to share more details! 🚀

**Comment by bhava07** [comment]
Thank you for getting back to me . I understand the issue with post-processing. I have worked with object detection models where the outputs are bounding boxes , classes and scores or big boxes, small boxes etc . i got the output shape for the said model is array([   1,   84, 8400]  . Could you please clarify on how to breakdown this array so as to get classes, bounding box co-ordinates and scores. I tried the code snippet that u have attached above but couldn't really provide clarity regarding the scores and classes . could you please share how to extract the classes boxes and scores for this scenario. 
thank you

**Comment by glenn-jocher** [comment]
@bhava07 Hi there! 🌟 It looks like you're on the right track with understanding the output structure of the TFLite model. Given the output shape `[1, 84, 8400]`, it seems like each detection is represented by 84 attributes, and you have 8400 such detections in total.

To break it down:
- The `84` likely includes the bounding box coordinates, confidence score, and class probabilities. 
- The exact breakdown within these 84 elements can vary based on the model, but typically, the first few elements represent the bounding box (e.g., `[x_center, y_center, width, height]`), followed by the confidence score, and then the class probabilities.

Without the specific model details, I'm making an educated guess here. So, your first step would be to understand or find the documentation on how your model outputs are structured.

Assuming the first 4 elements are for the bounding box, the 5th element is the confidence score, and the rest are class probabilities, you might extract the information like so:
```python
output_data = interpreter.get_tensor(output_details[0]['index'])

# Let's say your outputs are in 'output_data' with shape [1, 84, 8400]
detections = output_data[0]  # Removing the batch dimension

# Parsing each detection
boxes = detections[:, :4]  # First 4 elements for each detection
confidences = detections[:, 4]  # 5th element for confidence score
class_probs = detections[:, 5:]  # remaining elements for class probabilities

# Calculate class IDs
classes = np.argmax(class_probs, axis=1)

# Example to filter detections
valid_detections_idx = confidences > some_confidence_threshold
filtered_boxes = boxes[valid_detections_idx]
filtered_classes = classes[valid_detections_idx]
filtered_confidences = confidences[valid_detections_idx]
```
This is a simplified example to demonstrate the concept. You'd need to fine-tune it based on your actual model output structure.

Regarding the issue with all classes showing as 0, it could be due to model quantization or differences in input preprocessing that affect how the model perceives and classifies objects. Double-checking these aspects might help identify the root cause.

Hope this hints you in the right direction! If there's anything more specific or another area you'd like to dive deeper into, feel free to reach out. Keep experimenting! 😊
